packages/zoe/src/contracts/vpool-xyk-amm/pool.js

```diff
@@ -0,0 +1,255 @@
+// @ts-check
+
+import { E } from '@agoric/eventual-send';
+import { assert, details as X } from '@agoric/assert';
+import { AssetKind, AmountMath, isNatValue } from '@agoric/ertp';
+import { makeNotifierKit } from '@agoric/notifier';
+import { Far } from '@agoric/marshal';
+
+import {
+  calcLiqValueToMint,
+  calcValueToRemove,
+  calcSecondaryRequired,
+} from '../../contractSupport/index.js';
+
+import '../../../exported.js';
+import { makePriceAuthority } from '../multipoolAutoswap/priceAuthority.js';
+import { makeSinglePool } from './singlePool';
+
+// Pools represent a single pool of liquidity. Price calculations and trading
+// happen in a wrapper class that knows whether the proposed trade involves a
+// single pool or multiple hops.
+
+/**
+ * @param {ContractFacet} zcf
+ * @param {(brand: Brand) => boolean} isSecondary
+ * @param {(brand: Brand, pool: XYKPool) => void} initPool
+ * @param {Brand} centralBrand
+ * @param {Timer} timer
+ * @param {IssuerKit} quoteIssuerKit
+ * @param {bigint} protocolFee - Ratio, soon to be replaced with governed value
+ * @param {bigint} poolFee - Ratio, soon to be replaced with governed value
+ * @param {ZCFSeat} protocolSeat
+ */
+export const makeAddPool = (
+  zcf,
+  isSecondary,
+  initPool,
+  centralBrand,
+  timer,
+  quoteIssuerKit,
+  protocolFee,
+  poolFee,
+  protocolSeat,
+) => {
+  const makePool = (liquidityZcfMint, poolSeat, secondaryBrand) => {
+    let liqTokenSupply = 0n;
+
+    const {
+      brand: liquidityBrand,
+      issuer: liquidityIssuer,
+    } = liquidityZcfMint.getIssuerRecord();
+    const { notifier, updater } = makeNotifierKit();
+
+    const updateState = pool =>
+      // TODO: when governance can change the interest rate, include it here
+      updater.updateState({
+        central: pool.getCentralAmount(),
+        secondary: pool.getSecondaryAmount(),
+      });
+
+    const addLiquidityActual = (pool, zcfSeat, secondaryAmount) => {
+      const liquidityValueOut = calcLiqValueToMint(
+        liqTokenSupply,
+        zcfSeat.getAmountAllocated('Central').value,
```

Comment:
How do we know this value is a `Nat`? I don't think it's statically evident from just this module. `calcLiqValueToMint` will throw if it is not.

I don't see a defensive consistency risk (throwing fails safe) but as we learned in the testnet, just failing safe isn't enough for core economy components. Availability is nearly as important, so I think I should be checking for defensive _correctness_.

I can imagine that the argument for correctness in the face of a `zcfSeat` with a non-fungible `Central` allocation is straightforward, but I'd like us to agree on what it is and perhaps document it.

If the argument is that we know who all the callers are because they're inside this vat, that is coherent, but it means that the argument for correctness of this module isn't evident from this module alone, which is... messy.

Reply:
addLiquidity can't be called until the pool is created. `addPool)` in `pool.js` asserts that the `assetKind === AssetKind.NAT`.

Is that the kind of hazard that bit us in testNet?

Reply:
the thing that bit us in testnet wasn't just a hazard that made code hard to read; it was an actual bug in the runtime behavior of the software.

This comment addresses my concern.